src/include/binder/binder_context.h

```diff
@@ -0,0 +1,192 @@
+   * @param schema Schema object
+   * @param col_name Name of the column
+   * @return true if the column is in the schema, false otherwise
+   */
+  static bool ColumnInSchema(const catalog::Schema &schema, const std::string &col_name);
+
+  /**
+   * Construct the column position tuple given column name and the corresponding column value expression.
+   * Note that this is just a helper function and it is independent of the context.
+   * @param col_name Name of the column
+   * @param tuple Tuple of database oid, table oid, and schema object
+   * @param expr Column value expression
+   */
+  static void GetColumnPosTuple(const std::string &col_name,
+                                std::tuple<catalog::db_oid_t, catalog::table_oid_t, catalog::Schema> tuple,
+                                parser::ColumnValueExpression *expr);
+
+  /**
+   * Construct the column position tuple given only the column value expression and the context.
+   * Also internally update the column value expression according to the values in the context
+   * @param current_context Pointer to the current binder context
+   * @param expr Column value expression
+   * @return Returns true if the column is found in the alias maps of the current context; false otherwise
+   */
+  static bool GetColumnPosTuple(BinderContext *current_context, parser::ColumnValueExpression *expr);
+
+  /**
+   * Check if the table alias can be found in the alias maps of the current context or the upper contexts.
+   * This function internally updates the depth of the expression if the alias is successfully found
+   * @param current_context Current binder context
+   * @param alias Table alias
+   * @param expr Column value expression
+   * @param tuple Tuple of database oid, table oid, and schema object
+   * @return Return true if the alias is found, false otherwise
+   */
+  static bool GetRegularTableObj(BinderContext *current_context, const std::string &alias,
+                                 parser::ColumnValueExpression *expr,
+                                 std::tuple<catalog::db_oid_t, catalog::table_oid_t, catalog::Schema> *tuple);
+
+  /**
+   * Check if the table, represented by the table alias, has the column indicated by the column name.
+   * This function internally updates the information of the expression if the column is successfully found
+   * @param current_context Current binder context
+   * @param alias Table alias
+   * @param col_name Name of the column
+   * @param expr Column value expression
+   * @return Return true if the column is found, false otherwise
+   */
+  static bool CheckNestedTableColumn(BinderContext *current_context, const std::string &alias,
+                                     const std::string &col_name, parser::ColumnValueExpression *expr);
+
+  /**
+   * Get the pointer to the upper context of the current context
+   * @return Pointer to the upper binder context
+   */
+  BinderContext *GetUpperContext() { return upper_context_; }
```

Comment:
in last few lines of function `BindNodeVisitor::Visit(parser::SelectStatement *node)`, the `context_` field of the `BindNodeVisitor` object will be assigned with the upper_context; it may get modified in the future